Boot firmware updates support for ZynqMP

igoropaniuk opened the pull request:
Add command for retrieving/adjusting multiboot register value, which
permits to force CSU ROM t`o boot image from particular offset after
soft reset.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

Please do not submit a Pull Request via github.  Our project makes use of
mailing lists for patch submission and review.  For more details please
see https://www.denx.de/wiki/U-Boot/Patches

igoropaniuk committed:
[FIO toup] arm64: zynqmp: add cmd for adjusting multiboot register

Add command for retrieving/adjusting multiboot register value, which
permits to force CSU ROM t`o boot image from particular offset after
soft reset.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

ldts commented:
hey, better pick what is upstream (https://www.mail-archive.com/u-boot@lists.denx.de/msg413289.html) and then update accordingly

MrCry0 commented on board/xilinx/zynqmp/zynqmp.c:
```diff
@@ -344,22 +344,68 @@ int board_early_init_f(void)
+	"Get/Set CSU multiboot register",
+	"\n"
+	"   no param  - get current offset value\n"
+	"   [0xoffset] - set offset of the boot image\n"
```
I've never seen such a form of the parameter. Wouldn't it be better to use more usual [offset]?
Like in "ubi part [part] [offset]".

*_strto*() understand 8,10,16-base input values, so it's not needed to bind the input format as a hex.

ldts replied:
the offset defines 32KB blocks (ie, 1 = 32KB, 2 = 64K and so on.....maybe worth mentioning to avoid nasty surprises). Because of this,  I also dont think it makes sense to define it in hex....maybe better just decimal?

igoropaniuk replied:
Fixed

MrCry0 commented on board/xilinx/zynqmp/zynqmp.c:
```diff
@@ -344,22 +344,68 @@ int board_early_init_f(void)
+	if (argc == 1) {
+		multiboot = multi_boot_get();
+
+		printf("Multiboot:\t%d\n", multiboot);
```
Could you use hex format for all outputs of multiboot value? I think it would be more natural and readable.

igoropaniuk replied:
As we discussed below in the comments, the offset defines 32KB blocks, so it's better to print it as decimal value

ldts commented on board/xilinx/zynqmp/zynqmp.c:
```diff
@@ -344,22 +344,68 @@ int board_early_init_f(void)
+		return CMD_RET_SUCCESS;
+	}
+
+	multiboot = simple_strtoul(argv[1], NULL, 0);
```
why not use strict_strtoul and handle the error? this seems a delicate path so the more checks the better

igoropaniuk replied:
Fixed

igoropaniuk commented:
@MrCry0 @ldts all comments addressed

MrCry0 reviewed (approved):
LGTM

Commit:
[FIO fromlist] xilinx: zynqmp: use zynqmp_mmio_read() in multi_boot()

When U-Boot runs in EL2 there is no access to csu_base registers that's why
this has to be done via firmware interface to find out multi boot register
value. Till now this function is called only from SPL in EL3.

Signed-off-by: Michal Simek <michal.simek@xilinx.com>

igoropaniuk commented:
@ldts do you have any additional comments/suggestions?

igoropaniuk commented:
Decided to rename this PR and keep it open, I'll constantly push new commits related to boot firmware updates support

igoropaniuk committed:
[FIO toup] spl: mmc: support dynamic spl payload name on fs

Add weak function for detecting U-Boot SPL payload, this introduces
possibility to detect SPL payload in runtime (usually handy on A/B setups).

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO toup] zynqmp: detect spl payload based on multiboot offset value

Detect a filename of SPL payload based on the value of multiboot offset.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

MrCry0 reviewed (approved):
LGTM

ricardosalveti reviewed (approved):
LGTM

igoropaniuk committed:
[FIO toup] arm64: zynqmp: add cmd for getting boot auth state

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk commented:
@MrCry0 @ldts @ricardosalveti pushed additional commit, which implements U-Boot cmd for getting boot auth state

ricardosalveti commented:
> @MrCry0 @ldts @ricardosalveti pushed additional commit, which implements U-Boot cmd for getting boot auth state

Sure, let us know once you're 'done' with this PR for a final review.

igoropaniuk commented:
@ricardosalveti that's it, I don't plan to push anything else

ricardosalveti commented:
LGTM
